### DISCUSSION TITLE
SDK: Provide a way to disable the cache

### DISCUSSION THREAD

**Discussion by stondini** [directus/directus#22941] [Feature Requests]
### Summary

An external application imports data to Directus (10). Firstly, it fetches data from a legacy system and compare to the existing records in Directus. Secondly, it upsets or deletes records according to the comparison result.
The application must be sure that the fetched records from Direscrus are up to date. Meaning, the cache must be disabled in order to get the latest state of the records.

Also, the application authenticates using a user which is not an admin in order to avoid any other data manipulation.
That means, the application can't use the `clearCache()` command as only admin can.

### Motivation

Currently, in Directus 10, the only way to get the most recent data is to execute the `clearCache()` before each read command.
Also, the user must be admin of the whole system which is not secure.

### Detailed Design

A possible implementation could be to add a `disableCache()` extension like below in TypeScript:
```
    const client = createDirectus(this.config.url.toString())
      .with(staticToken(this.config.accessToken))
      .with(rest())
      .with(disableCache());
      
    client.request(...) // Fetch the up to date and committed data.
```

### Must Have:
- An option to disable the cache (either globally to the client instance or per request.

### Should Have:
- N/A

### Could Have:
- N/A

### Won't Have:
- N/A

### Drawbacks

The "disable cache" option must not impact the other clients of the API.

### Alternatives

Instead of exposing a new `disableCache` extension, a per-request option could also work which offers a fined grained cache control.

Example below in TypeScript:
```
  const result = .client
      .request<TenantDto[]>(
        readItems('articles', {
          fields: ['*'],
          filter: {
            status: {
              _eq: 'published',
            }
          },
          limit: 10,
          disableCache: true // default false
        }),
      )
```

### Adoption Strategy

Ir should not be a breaking change as the default behaviour remains the same. Only explicit configuration disables the cache.

**Comment by br41nslug** [comment]
The SDK doesnt do any caching for you to disable 🤔 You can disable browser cache by poviding a `Cache-Control: no-store` header (if Directus is configured to allow this less secure option). Beyond that you'll have to disable/configure caching on the Directus end or call the cache clear endpoint, both approaches require admin privileges as allowing non-admin users to bypass configured cache is a security/performance issue.

**Comment by domdus** [comment]
Take a look at `CACHE_AUTO_PURGE` setting.

https://docs.directus.io/self-hosted/config-options.html#cache
[2] CACHE_AUTO_PURGE allows you to keep the Directus API real-time, while still getting the performance benefits on quick subsequent reads.

**Reply by br41nslug** [reply]
That would fall under "configuring cache on the Directus end", you can configure this environment variable as desired if you're the admin of the instance but thats unrelated to the API/SDK as they dont have access to such backend configuration.

**Comment by stondini** [comment]
Thank you for your comments and suggestion.

I enabled `CACHE_SKIP_ALLOWED` option and I'm trying to pass the `Cache-Control` header.
Do you know how to pass such header in the following statement ?
```
await this.client.request<T[]>(
      readItems(collectionName, {
        fields: itemFields,
        filter: {
          tenant: {
            _eq: this.tenant.id,
          },
        },
        limit: -1,
      }),
    );
```

**Reply by domdus** [reply]
Below is an example on how to do it on directus client init.
(Found here: https://docs.directus.io/guides/headless-cms/build-static-website/next.html#create-a-helper-for-the-sdk)

It affects the client fetch API cache at least. I don't know if that's the way to impact the directus (server side) API cache as well - try it.

```
const directus = createDirectus('https://directus.example.com').with(
  rest({
    onRequest: (options) => ({ ...options, cache: 'no-store' }),
  })
);
```

**Reply by stondini** [reply]
Thank you @domdus for your suggestion.

I tried and checked using Wireshark and, `Cache-Control` header is sent with every request:
```
GET /items/contact_channels  HTTP/1.1
host: localhost:8055
connection: keep-alive
Content-Type: application/json
Authorization: Bearer <MASKED>
accept: */*
accept-language: *
sec-fetch-mode: cors
user-agent: node
pragma: no-cache
cache-control: no-cache
accept-encoding: gzip, deflate
```

But unfortunately, it doesn't work.
The application still gets `RECORD_NOT_UNIQUE` error.

Just to recap how the application works:

1. Fetches data from a legacy system.
2. Fetches data from Directus.
3. Compares both data sets and upset in Directus.

During the first run, Directus DB is empty so, all records are inserted.
Then, on the second run (a few **seconds** later), the step 2 returns an empty data set as Directus cache is empty. It makes the application re-inserting all the records.
When I run the application a few **minutes** later, the step 2 returns the full data set as Directus cache is full. It makes the application doing nothing as all records are the same.

It works well if the application calls `clearCache()` endpoint but I would prefer to not grant the Admin role to the app.

**Reply by domdus** [reply]
@stondini from your example request above: `cache-control: no-cache`

Make sure the header value is `no-store` (not `no-cache`). That should do the trick.

``cache-control: no-store``

**Reply by stondini** [reply]
Thank you @domdus 

I've just tries and it works.
There is a behavioural difference between:
```
const directus = createDirectus('https://directus.example.com').with(
  rest({
    onRequest: (options) => ({ ...options, cache: 'no-store' }),
  })
);
```
and
```
const directus = createDirectus('https://directus.example.com').with(
  rest({
        onRequest: (options) => ({
          ...options,
          headers: {
            ...options.headers,
            'cache-control': 'no-store',
          },
        }),
      })
);
```

Anyway, the explicit `cache-control` header in addition to the server `CACHE_SKIP_ALLOWED=true` setting works well.
I don't need to assign `Admin` role to the user the app authenticates with.

Thank you.
